doc/decisions/DR023-Versioning-traits-and-specifications.md

```diff
@@ -0,0 +1,702 @@
+
+
+# DR023 Versioning traits and specifications
+
+- **Status:** To be decided
+- **Impact:** High
+- **Driver:** @elliotcmorris
+- **Approver:** @feltech
+- **Contributors:** @elliotcmorris @feltech @reinecke @antirotor
+- **Outcome:**  TBC
+
+### Preface
+This document refers to the versioning of trait and specification
+libraries in general. However, as `openassetio-mediacreation` is the
+only one in play at the moment, and since reading "trait and
+specifications library" over and over would be tiresome, we will just
+use the term `mediacreation` throughout this document.
+
+# Problem Statement
+In order to achieve OpenAssetIOs interoperability goal, hosts and
+managers must agree on the structure of data exchanged between them.
+
+The `openassetio-mediacreation` defines the agreed standard for interop
+objects (called `traits` and `specifications`) across the VFX industry.
+Hosts and Managers must agree on the structure of these objects. [These
+objects](https://github.com/OpenAssetIO/OpenAssetIO-MediaCreation/blob/main/traits.yml)
+are currently in a first-pass state, and it's understood that the only
+way to come to a solid set of useful shared concepts is via use and
+iteration.
+
+OpenAssetIO wants to be thin, and does not want to be "the problem". To
+this end, we need a change and versioning process for the traits and
+specifications schema which allows maximum compatibility, especially in
+the face of the unpredictable version adoption in the VFX industry.
+
+[`openassetio-mediacreation`](https://github.com/OpenAssetIO/OpenAssetIO-MediaCreation)
+is a generative library, in that it produces trait and specification
+view classes representing the objects specified in the source `.yml` via
+the `openassetio-traitgen` tool, currently in both the c++ and python
+languages. As such, the challenge of versioning is not only descriptive,
+but also technical.
+
+# Aims :
+- Allow new mediacreation versions to be adopted by managers without
```

Comment:
This raises the spectre of some kind of negotiation between host and manager, and requiring _both_ hosts and managers to bloat their codebase with lots of conditionals, rather than isolating that bloat to just the (typically more nimble) manager implementers.

I'm sure pipeline people in the past have told us that they're used to new versions of DCC tools breaking their plugins, and have processes in place to adapt when upgrading. Not to mention the cadence of adopting new DCC tool releases tends to be slow enough that they'd have plenty of warning of breaking changes (and there would likely be  other breaking changes besides OpenAssetIO).

Commercial plugins are an interesting case, though I would hope they'd be self-motivated to keep up to date.

Having said all that, I'm not totally averse to a negotiation mechanism, I'm just keen to limit the scope of a versioning mechanism to the minimum that would be generally helpful. If trait version negotiation is really required for that, then so be it.

Reply:
Yeah, I think a "negotiation mechanism" (which includes one way demands) are my main concern, as it implies that one or more parties have to understand _all_ data. The original API design was that "managers faithfully store and recall data, unless in specific circumstances they _do_ have first-class understanding (eg paths)" and that case was usually in the minority. Any kind of negotiation mechanism significantly affects the viability of that. The data being self describing however still facilitates this and places the burden on whoever wishes to specifically _interpret_ the data.

Reply:
> managers faithfully store and recall data, unless in specific circumstances they do have first-class understanding

To be clear, that's only for managers that advertise such a capability (via `managementPolicy`), not all managers.

To make sure I understand the wider point - a naive "pass-through" store-and-recall manager would never be able to respond positively to a negotiation, since it has no opinion on what trait/schema version is used?

I guess similarly, a generic host (e.g. `simpleResolver`) would have the same problem, if it was expected that the host had to provide a schema version to the manager, when it has no opinion.

Reply:
> To make sure I understand the wider point - a naive "pass-through" store-and-recall manager would never be able to respond positively to a negotiation, since it has no opinion on what trait/schema version is used?

sorry missed that bit - yeah, thats my main concern with any active negotiation mechanism.

Reply:
> Could it still work, though, as long as there is a way for the host/manager to signal in its negotiation "I don't care" - perhaps that's the default implementation.

Visits a previous question - is versioning optional? This negotiation mechanism sounds complex in practice and requires a lot of extra code, if its not part of how you handle data inherently, I bet most people would skip it- as its hard enough as is to understand an integration as a newcomer?

Reply:
> I bet most people would skip it- as its hard enough as is to understand an integration as a newcomer?
I tend to agree. Openassetio has already broken the seal on requiring a certain level of manual diligence on the implementors part, "good behaviour", etc, but it does seem like negotiated would be a humongous addition to that. If an openassetio integration isn't already impossible without a reference currently, it definitely would be under that paradigm.